File: LeetCode/0036/golang/lc0036.go
package lc0036

// Runtime complexity: O(9*9) == O(1)
// Auxiliary space: O(3*9*9) == O(1)
// Subjective level: medium
func isValidSudoku(board [][]byte) bool {
	rows := make([][]bool, 9)
	for k := range 9 {
		rows[k] = make([]bool, 9)
	}
	cols := make([][]bool, 9)
	for k := range 9 {
		cols[k] = make([]bool, 9)
	}
	grids := make([][]bool, 9)
	for k := range 9 {
		grids[k] = make([]bool, 9)
	}

	for h := range 9 {
		for w := range 9 {
			if board[h][w] == '.' {
				continue
			}
			// at this point board[h][w] is guaranteed to be a byte between '1' and '9'
			num := int(board[h][w] - '1') // num is in [0, 8] range.

			if rows[h][num] {
				return false
			}
			rows[h][num] = true

			if cols[w][num] {
				return false
			}
			cols[w][num] = true

			gridnum := 3*(h/3) + (w / 3)
			if grids[gridnum][num] {
				return false
			}
			grids[gridnum][num] = true
		}
	}

	return true
}

File: LeetCode/0036/python/lc0036.py
from typing import List

# Runtime complexity: O(9*9) == O(1)
# Auxiliary space complexity: O(3*9*9) == O(1)
# Subjective level: medium leaning into easy.
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [[False]*9 for row in range(9)]
        cols = [[False]*9 for row in range(9)]
        grids = [[False]*9 for row in range(9)]

        for h in range(9):
            for w in range(9):
                if board[h][w] == '.':
                    continue
                num = int(board[h][w]) - 1

                if rows[h][num]:
                    return False
                rows[h][num] = True

                if cols[w][num]:
                    return False
                cols[w][num] = True

                gridnum = 3 * (h//3) + w//3
                if grids[gridnum][num]:
                    return False
                grids[gridnum][num] = True
        return True
